File: test/samples/vulnerable.java
// SQL Injection in Java
public class VulnerableJava {
    
    public List<User> getUserById(String userId) {
        String query = "SELECT * FROM users WHERE id = '" + userId + "'";
        return jdbcTemplate.query(query, new UserRowMapper());
    }
    
    // XSS Vulnerability
    public void displayComment(HttpServletResponse response, String comment) throws IOException {
        response.getWriter().println("<div class='comment'>" + comment + "</div>");
    }
    
    // Command Injection
    public String executeCommand(String userInput) throws IOException {
        Runtime runtime = Runtime.getRuntime();
        Process process = runtime.exec("ls " + userInput);
        return "Command executed";
    }
    
    // Path Traversal
    public String readFile(String fileName) throws IOException {
        File file = new File("./user_files/" + fileName);
        return new String(Files.readAllBytes(file.toPath()));
    }
    
    // Weak Cryptography
    public String hashPassword(String password) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] digest = md.digest(password.getBytes());
            return new BigInteger(1, digest).toString(16);
        } catch (NoSuchAlgorithmException e) {
            return null;
        }
    }
    
    // Hard-coded Credentials
    public Connection connectToDatabase() throws SQLException {
        return DriverManager.getConnection(
            "jdbc:mysql://localhost:3306/mydb", 
            "admin", 
            "password123"
        );
    }
    
    // Insecure Random
    public String generateToken() {
        return "token_" + Math.random();
    }
    
    // XXE Vulnerability
    public Document parseXml(String xml) throws Exception {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        return builder.parse(new InputSource(new StringReader(xml)));
    }
    
    // Insufficient Logging
    public void login(String username, String password) {
        // No logging of login attempts
        if (authenticate(username, password)) {
            grantAccess();
        }
    }
    
    // Insecure File Upload
    public void saveFile(InputStream fileStream, String fileName) throws IOException {
        // No validation of file extension or content
        Files.copy(fileStream, Paths.get("/uploads/" + fileName));
    }
    
    // Unvalidated Redirect
    public void redirect(HttpServletResponse response, String url) throws IOException {
        response.sendRedirect(url);
    }
    
    // LDAP Injection
    public List<User> searchUsers(String username) throws NamingException {
        DirContext context = new InitialDirContext();
        String searchFilter = "(uid=" + username + ")";
        NamingEnumeration<SearchResult> results = context.search("ou=users,dc=example,dc=com", searchFilter, null);
        // Process results
        return new ArrayList<>();
    }
} 

File: test/samples/vulnerable.js
// SQL Injection
function getUserData(userId) {
  const query = "SELECT * FROM users WHERE id = '" + userId + "'";
  return db.query(query);
}

// XSS
function displayUserComment(comment) {
  document.getElementById('comments').innerHTML = comment;
}

// Command Injection
function executeCommand(userInput) {
  const { exec } = require('child_process');
  exec('ls ' + userInput, (error, stdout, stderr) => {
    console.log(stdout);
  });
}

// Insecure JWT
function verifyToken(token) {
  const jwt = require('jsonwebtoken');
  return jwt.verify(token, secretKey);  // Missing algorithm specification
}

// Path Traversal
function readUserFile(fileName) {
  const fs = require('fs');
  return fs.readFileSync('./user_files/' + fileName);
}

// Weak Cryptography
function hashPassword(password) {
  const crypto = require('crypto');
  return crypto.createHash('md5').update(password).digest('hex');
}

// Prototype Pollution
function mergeUserData(userData) {
  const target = {};
  function merge(target, source) {
    for (let key in source) {
      target[key] = source[key];
    }
    return target;
  }
  return merge(target, userData);
}

// Insecure Random Values
function generateToken() {
  return 'token_' + Math.random().toString(36).substring(2);
}

// NoSQL Injection
function findUser(username) {
  return db.collection.find({ username: username });
}

// Insecure Deserialization
function processUserData(userData) {
  return eval('(' + userData + ')');
}

// CSRF vulnerability
function setupForm() {
  app.post('/update-profile', (req, res) => {
    // No CSRF token validation
    updateUserProfile(req.body);
    res.redirect('/profile');
  });
}

// SSRF vulnerability
function fetchExternalResource(url) {
  const axios = require('axios');
  return axios.get(url);
} 